### PR TITLE
PID Based Drive

### LINKED ISSUE #27: PID based drive
*Voltage

### PR DESCRIPTION
**Pull request by ParkerMeyers**
PID Based Drive instead of percentage based drive

Closes #27

### CODE REVIEW

File: src/main/java/bhs/devilbotz/Constants.java [hunk 1]
```diff
@@ -17,11 +17,32 @@ public final class Constants {
...
+
+    // Drive PID values TODO: Tune
+    public static final double driveP = 3.5725;
+    public static final double driveI = 0.0;
+    public static final double driveD = 0.0;
+
+    // Feedforward constants TODO: Tune
+    public static final double driveFFS = 0.94143;
+    public static final double driveFFV = 2.3803;
+    public static final double driveFFA = 0.48128;
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/Constants.java]: Hi Parker, you need to change these to google style constants.  E.g. max_speed --> MAX_SPEED

File: src/main/java/bhs/devilbotz/commands/DriveCommand.java [hunk 2]
```diff
@@ -8,16 +9,16 @@
...
   private final DriveTrain drive;
   private final DoubleSupplier speed;
-  private final DoubleSupplier rotation;
+  private final DoubleSupplier rot;
 
-  private final SlewRateLimiter speedSlewRateLimiter = new SlewRateLimiter(3);
+  private final SlewRateLimiter speedSlewRateLimiter = new SlewRateLimiter(10);
 
-  private final SlewRateLimiter rotationSlewRateLimiter = new SlewRateLimiter(3);
+  private final SlewRateLimiter rotationSlewRateLimiter = new SlewRateLimiter(10);
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/commands/DriveCommand.java]: These should be defined in Constants.java

File: src/main/java/bhs/devilbotz/Robot.java [hunk 3]
```diff
@@ -34,6 +37,7 @@ public void robotInit() {
     // Instantiate our RobotContainer.  This will perform all our button bindings, and put our
     // autonomous chooser on the dashboard.
     robotContainer = new RobotContainer();
+    drive = robotContainer.getDriveTrain();
     shuffleboardManager = robotContainer.getShuffleboardManager();
   }
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/Robot.java:43]: We don't need this code.

File: src/main/java/bhs/devilbotz/RobotContainer.java [hunk 4]
```diff
@@ -74,4 +74,8 @@ public Command getAutonomousCommand(AutonomousModes autoMode) {
   public ShuffleboardManager getShuffleboardManager() {
     return shuffleboardManager;
   }
-}
\ No newline at end of file
+
+  public DriveTrain getDriveTrain() {
+    return driveTrain;
+  }
+}
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/RobotContainer.java:81]: We don't need this code and shouldn't expose the driveTrain object.

File: src/main/java/bhs/devilbotz/commands/DriveCommand.java [hunk 5]
```diff
@@ -8,16 +9,18 @@
 public class DriveCommand extends CommandBase {
   private final DriveTrain drive;
   private final DoubleSupplier speed;
-  private final DoubleSupplier rotation;
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/commands/DriveCommand.java:11]: Why not leave it as "rotation"?  It's easier to read/understand, but no big deal.

File: src/main/java/bhs/devilbotz/subsystems/DriveTrain.java [hunk 6]
```diff
@@ -66,22 +227,57 @@ public void resetNavx() {
...
   }
 
+  /**
+   * Updates the odometry of the drive train. This method is called in the periodic method.
+   *
+   * @since 1/30/2023
+   */
+  public void updateOdometry() {
+    odometry.update(navx.getRotation2d(), getLeftDistance(), getRightDistance());
+    field.setRobotPose(odometry.getPoseMeters());
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/subsystems/DriveTrain.java:272]: updateOdometry should be private

**Review by BrownGenius — CHANGES_REQUESTED** [review]
Please remove the unused functions and change the public methods to private.

File: src/main/java/bhs/devilbotz/Constants.java (hunk 1 above)
Comment (ParkerMeyers) [reply src/main/java/bhs/devilbotz/Constants.java]: Already done

File: src/main/java/bhs/devilbotz/commands/DriveCommand.java (hunk 2 above)
Comment (ParkerMeyers) [reply src/main/java/bhs/devilbotz/commands/DriveCommand.java]: Already done

File: src/main/java/bhs/devilbotz/commands/DriveCommand.java (hunk 5 above)
Comment (ParkerMeyers) [reply src/main/java/bhs/devilbotz/commands/DriveCommand.java:11]: I can put it back, but more WPILIB methods use rot instead of rotation

**Review by BrownGenius — APPROVED** [review]

**Pull request merged**